lib/theme_check/language_server/variable_lookup_finder.rb

```diff
@@ -0,0 +1,322 @@
+          \]|
+          # opening [
+          \[
+        )$
+      }x
+
+      VARIABLE_LOOKUP_CHARACTERS = /[a-z0-9_.'"\]\[]/i
+      VARIABLE_LOOKUP = /#{VARIABLE_LOOKUP_CHARACTERS}+/o
+      SYMBOLS_PRECEDING_POTENTIAL_LOOKUPS = %r{
+        (?:
+          \s(?:
+            if|elsif|unless|and|or|#{Liquid::Condition.operators.keys.join("|")}
+            |echo
+            |case|when
+            |cycle
+            |in
+          )
+          |[:,=]
+        )
+        \s+
+      }omix
+      ENDS_WITH_BLANK_POTENTIAL_LOOKUP = /#{SYMBOLS_PRECEDING_POTENTIAL_LOOKUPS}$/oimx
+      ENDS_WITH_POTENTIAL_LOOKUP = /#{SYMBOLS_PRECEDING_POTENTIAL_LOOKUPS}#{VARIABLE_LOOKUP}$/oimx
+
+      def lookup(content, cursor)
+        return if cursor_is_on_bracket_position_that_cant_be_completed(content, cursor)
+        potential_lookup = lookup_liquid_variable(content, cursor) || lookup_liquid_tag(content, cursor)
+
+        # And we only return it if it's parsed by Liquid as VariableLookup
+        return unless potential_lookup.is_a?(Liquid::VariableLookup)
+        potential_lookup
+      end
+
+      private
+
+      def cursor_is_on_bracket_position_that_cant_be_completed(content, cursor)
+        content[0..cursor - 1] =~ ENDS_IN_BRACKET_POSITION_THAT_CANT_BE_COMPLETED
+      end
+
+      def cursor_is_on_liquid_variable_lookup_position(content, cursor)
+        previous_char = content[cursor - 1]
+        is_liquid_variable = content =~ Liquid::VariableStart
+        is_in_variable_segment = previous_char =~ VARIABLE_LOOKUP_CHARACTERS
+        is_on_blank_variable_lookup_position = content[0..cursor - 1] =~ /[{:,-]\s+$/
+        (
+          is_liquid_variable && (
+            is_in_variable_segment ||
+            is_on_blank_variable_lookup_position
+          )
+        )
+      end
+
+      def lookup_liquid_variable(content, cursor)
+        return unless cursor_is_on_liquid_variable_lookup_position(content, cursor)
+        start_index = content.match(/#{Liquid::VariableStart}-?/o).end(0) + 1
+        end_index = cursor - 1
+
+        # We take the following content
+        # - start after the first two {{
+        # - end at cursor position
+        #
+        # That way, we'll have a partial liquid variable that
+        # can be parsed such that the "last" variable_lookup
+        # will be the one we're trying to complete.
+        markup = content[start_index..end_index]
+
+        # Early return for incomplete variables
+        return empty_lookup if markup =~ /\s+$/
+
+        # Now we go to hack city... The cursor might be in the middle
+        # of a string/square bracket lookup. We need to close those
+        # otherwise the variable parse won't work.
+        markup += "'" if markup.count("'").odd?
+        markup += '"' if markup.count('"').odd?
+        markup += "]" if markup =~ UNCLOSED_SQUARE_BRACKET
+
+        variable = variable_from_markup(markup)
+
+        variable_lookup_for_liquid_variable(variable)
```

Comment:
Is this logic that different than the one used by the tag? I'm using Variable.new directly but I think that if I didn't strip the start and closed it appropriately, I could have all the logic in one.

Reply:
It depends where in tags and what tag. For the `if` conditional for eg. it's not the same. It's a `Condition` IIRC.